app/controllers/physical_server_controller.rb

```diff
@@ -36,4 +36,68 @@ def textual_group_list
     []
   end
   helper_method :textual_group_list
+
+  # Handles buttons pressed on the toolbar
+  def button
+    # Get the list of servers to apply the action to
+    servers = retrieve_servers
+
+    # Apply the action (depending on the button pressed) to the servers
+    apply_action_to_servers(servers)
+  end
+
+  private
+
+  # Maps button actions to actual method names to be called and the
+  # corresponding result status messages to be displayed
+  ACTIONS = {"physical_server_power_on"         => [:power_on,         _("Power On")],
+             "physical_server_power_off"        => [:power_off,        _("Power Off")],
+             "physical_server_restart"          => [:restart,          _("Restart")],
+             "physical_server_blink_loc_led"    => [:blink_loc_led,    _("Blink LED")],
+             "physical_server_turn_on_loc_led"  => [:turn_on_loc_led,  _("Turn On LED")],
+             "physical_server_turn_off_loc_led" => [:turn_off_loc_led, _("Turn Off LED")]}.freeze
+
+  # Returns a list of servers to which the button action will be applied
+  def retrieve_servers
+    server_id = params[:id]
+    servers = []
+
+    # A list of servers
+    if @lastaction == "show_list"
+      server_ids = find_checked_items
+      server_ids.each do |id|
+        servers.push(PhysicalServer.find_by('id' => id))
+      end
+      if server_ids.empty?
+        display_error_message("No server IDs found for the selected servers")
+      end
+    # A single server
+    elsif server_id.nil? || PhysicalServer.find_by('id' => server_id).nil?
+      display_error_message("No server ID found for the current server")
+    else
+      servers.push(PhysicalServer.find_by('id' => server_id))
+    end
+
+    servers
+  end
+
+  # Applies the appropriate action to a list of servers depending
+  # on the button pressed
+  def apply_action_to_servers(servers)
+    button_pressed = params[:pressed]
+
+    # Apply the appropriate action to each server
+    if ACTIONS.key?(button_pressed)
+      method = ACTIONS[button_pressed][0]
+      action_str = ACTIONS[button_pressed][1]
+      servers.each do |server|
+        server.send(method)
```

Comment:
Don't forget, that all actions that actually do things need to be caller throught the queue.

The front-end server (UI appliance) need not see the backend that has access to whatever API you are calling.

So the pattern to follow is that the UI code just puts tasks on the queue.

Better yet use the manageiq REST API to fire the actions.